### PR TITLE
[AI] [Micro AI] [1.17] Add new Micro AI: ai_type=bloodlust

### LINKED ISSUE #6372: [1.17/1.18] [FR] [Micro AI] Please add "scenario_boss" and "minion" Micro AI types for scenario/dungeon bosses.
### Describe the desired feature
Normally, the hunter MAI would work for this completely however, it has one distinct flaw: the return to nest behavior and rest until full heal thing. Thus, what I request is a subset of this MAI type as `scenario_boss` which would work as the hunter MAI but not retreat to a nest. I do have something like this made in my works so I can get those files in after cleaning the code.

Additionally, a `minion` MAI type can added in as well to mimic behavior for units which might spawn support units with the boss-class unit. Previously, the minions MAI was doable with a combination of lurker and goto MAI as shown as "Undead Empire" UMC. However, I think it's convenient to just make one micro AI for this instead of a combination of 4-5 MAI to achieve it via hard-to-read-code.

- [ ] 'scenario_boss' MAI
- [ ] 'minion' MAI

### Mainline Scenarios/Units which could benefit from this
1. Skeletal Dragon boss in EI (unless the campaign is reworked)
2. Xanthric the dragon in DiD, the three minibosses of Bones, Flesh and Spirit of DiD
3. dragon boss of WoF if WoF enters mainline 
4. Final boss of UtbS S12 if 6142 is achieved.
5. maybe some scenarios of upcoming mainline candidate campaigns. 

Additionally, many UMC campaigns will benefit, notably ones which have boss fights.

### PR DESCRIPTION
**Pull request by knyghtmare**
Here is a subset of the hunter MAI, called `ai_type=boss`, intended for dungeon and scenario boss-category units.

Sample Code:
```ini
[micro_ai]
        side=2
        ai_type=boss
        action=add

        id="Mal Keshar"
        [filter_location]
            x=30
            y=31
            radius=6
        [/filter_location]
[/micro_ai]
```

What it does?

1. Makes the 'boss' wander around a specified area (boss arena)
2. If enemy is within the boss's movement, target weakest enemy.
3. Approach and attack the weakest enemy.
4. Kill and reset to wandering mode.

Pending Stuff before making a full PR:

- [x] Test map
- [x] Test Scenario
- [ ] Documentation when merged.
- [x] Update Code to accomodate LuaAPI changes

Closes #6372

### CODE REVIEW

**Comment by stevecotton** [comment]
The name "boss" doesn't seem descriptive to me - I'd expect "boss" to be something like the standard AI's behavior for leaders, where if they're not on a castle they get on a castle, but if they are on a castle they might attack. I'd also suggest searching the forums and seeing if the name hits a lot of unrelated topics, currently "boss" finds 2152 existing posts.

Not sure if "gladiator" would be better, it's what the description in this PR suggests to me, but it's also a term that's been heavily used in the forums with 466 matching posts.

**Comment by knyghtmare** [comment]
> Not sure if "gladiator" would be better, it's what the description in this PR suggests to me, but it's also a term that's been heavily used in the forums with 466 matching posts.

TBH, the current behaviour is a hyper aggressive unit (or group of them) who will ruthlessly attack enemy units within the stated area. Possible tentative names:

- scenario_boss
- gladiator
- berserker
- aggressor
- hunter_no_home

**Comment by knyghtmare** [comment]
I have requested reviews from octalot (as the current AI project member is unfilled) and CelticMinstrel (as he's the expert on Lua)

File: data/ai/micro_ais/scenarios/dungeon_mai_comparisons.cfg
Comment (CelticMinstrel) [inline data/ai/micro_ais/scenarios/dungeon_mai_comparisons.cfg]: I'm guessing this would work if you add `data/` in front.

**Comment by CelticMinstrel** [comment]
I don't see any major problems with this. I haven't taken a close look at the actual AI code though.

File: data/ai/micro_ais/scenarios/dungeon_mai_comparisons.cfg
Comment (knyghtmare) [reply data/ai/micro_ais/scenarios/dungeon_mai_comparisons.cfg]: I rectified this when I found it was due to me not pushing the changes from my fork to the PR and then pulling to my local build

**Comment by CelticMinstrel** [comment]
Is the MAI abbreviation used in other test scenarios? I think it's better not to use it, but if others already do, then whatever.

**Comment by knyghtmare** [comment]
> Is the MAI abbreviation used in other test scenarios? I think it's better not to use it, but if others already do, then whatever.

Yeah, good point. I have to change the scenario id and name to something less technical, in like 2 words or less.

File: data/ai/micro_ais/scenarios/dungeon_mai_comparisons.cfg
Comment (stevecotton) [inline data/ai/micro_ais/scenarios/dungeon_mai_comparisons.cfg:283]: .. so as to ...

File: data/ai/micro_ais/scenarios/dungeon_mai_comparisons.cfg
Comment (stevecotton) [inline data/ai/micro_ais/scenarios/dungeon_mai_comparisons.cfg:289]: There's a pronoun mismatch in there ("he" as opposed to "it"). Trivial stuff, but might as well comment on it.

**Comment by stevecotton** [comment]
Looks reasonable to me, but the AI is an area that's unfamiliar to me.

Please add it to data/ai/micro_ais/scenarios/scenario_micro_ai.cfg.

File: data/ai/micro_ais/scenarios/dungeon_mai_comparisons.cfg
Comment (knyghtmare) [reply data/ai/micro_ais/scenarios/dungeon_mai_comparisons.cfg]: Fixed

File: data/ai/micro_ais/scenarios/dungeon_mai_comparisons.cfg
Comment (knyghtmare) [reply data/ai/micro_ais/scenarios/dungeon_mai_comparisons.cfg]: fixed.

**Comment by knyghtmare** [comment]
> Looks reasonable to me, but the AI is an area that's unfamiliar to me.
> 
> Please add it to data/ai/micro_ais/scenarios/scenario_micro_ai.cfg.

Okay, I have made the adjustments. What else is left to do before squash merging (or, I do need squash merging or not?).

**Comment by soliton-** [comment]
Commits should be squashed if they're not very useful history. Like typos introduced in one commit and fixed in a later one or pretty much any other instances of fixing up earlier commits. Or just plain bad commits like the "Changed file X" ones you get from website commits when you don't actually write a commit message. If none of those exist or were already squashed and force pushed there is no reason to further squash good commits.

**Comment by CelticMinstrel** [comment]
As far as I can tell, there's no good reason for this change to have 23 commits (most of them are fixups, reverts, or minor addressing feedback type of commits), so I'd just use the squash and merge button, which also means that @knyghtmare doesn't need to do any work to squash stuff.

**Comment by knyghtmare** [comment]
A squash and merge would be fine. Seems appropriate for this PR.

File: data/ai/micro_ais/scenarios/dungeon_mai_comparisons.cfg
Comment (stevecotton) [inline data/ai/micro_ais/scenarios/dungeon_mai_comparisons.cfg:38]: One of the reasons that the CI is failing is that wmlindent needs to be run on this file (or the whitespace fixed).

File: data/ai/micro_ais/scenarios/dungeon_mai_comparisons.cfg
Comment (stevecotton) [inline data/ai/micro_ais/scenarios/dungeon_mai_comparisons.cfg:58]: The schema is also missing `side`. I think it's easiest to just delete it from this file for now.

Comment (CelticMinstrel) [reply data/ai/micro_ais/scenarios/dungeon_mai_comparisons.cfg]: I think 0 is also the default?

**Comment by stevecotton** [comment]
> If enemy is within the defined area, target weakest enemy.

That doesn't seem to be what it does - the defined area only seems to affect the hexes that the boss may choose for random movement. The AI seems to ignore enemies within the defined area if the boss can't reach them in the current turn.

File: data/ai/micro_ais/maps/dungeon_mai_comparisons.map
Comment (stevecotton) [inline data/ai/micro_ais/maps/dungeon_mai_comparisons.map:6]: The map is decorative but doesn't give a good illustration of the bosses' areas. I assumed that each courtyard would be the bosses' area, but the areas are (radius 5) which the courtyards are (radius 6 with 6 corners removed).

Comment (knyghtmare) [reply data/ai/micro_ais/maps/dungeon_mai_comparisons.map:6]: The map was made like that to exhibit the case of the elite-class units behavior on units which are in the zone and within or outside their movement range.

**Comment by knyghtmare** [comment]
> > If enemy is within the defined area, target weakest enemy.
> 
> That doesn't seem to be what it does - the defined area only seems to affect the hexes that the boss may choose for random movement. The AI seems to ignore enemies within the defined area if the boss can't reach them in the current turn.

Oops. I mistyped that one. My bad. Rectified. Same for the in-scenario note/description.

**Comment by stevecotton** [comment]
The "will pursue and will not stop" sounds like it fixes on a victim, like The Terminator or various horror monsters. The actual behavior is that it'll return to its zone if the victim manages to run outside its movement zone.

Suggestion: rename it something like "sulking bloodlust". If any enemy is within this units' attack range, it will attack the weakest, even if this requires it to move away from its assigned zone. If no enemies are within range, it will wander randomly within its assigned zone, returning there if necessary.

File: data/ai/micro_ais/maps/dungeon_mai_comparisons.map
Comment (stevecotton) [reply data/ai/micro_ais/maps/dungeon_mai_comparisons.map:6]: I'm don't understand. The visual clues on the map are making me think the zones are different to their actual hexes.

What are the castle hexes around the zones meant to mean?

**Comment by knyghtmare** [comment]
> Suggestion: rename it something like "sulking bloodlust". If any enemy is within this units' attack range, it will attack the weakest, even if this requires it to move away from its assigned zone. If no enemies are within range, it will wander randomly within its assigned zone, returning there if necessary.

That works. Also, is it implied to rename it from `ai_type=boss` to `ai_type=bloodlust`?
The name sounds appropriate

File: data/ai/micro_ais/maps/dungeon_mai_comparisons.map
Comment (knyghtmare) [reply data/ai/micro_ais/maps/dungeon_mai_comparisons.map:6]: The castle hexes around the zones dont serve any role other than to further augment the appearance of an arena. Aesthetics.

Comment (stevecotton) [reply data/ai/micro_ais/maps/dungeon_mai_comparisons.map:6]: For the micro ai demo scenarios, I think understanding is more important than aesthetics. You could do with labels on the map, but please communicate the AI's zones to the person viewing the map.

File: data/ai/micro_ais/cas/ca_bloodlust.lua
Comment (stevecotton) [inline data/ai/micro_ais/cas/ca_bloodlust.lua:64]: A few of the comments have been hit by a find-and-replace, and should possible just say "unit" or similar.

**Review by stevecotton — APPROVED** [review]
Looks good. My only comment is a trivial issue, which you can decide to ignore and merge as-is.

**Comment by knyghtmare** [comment]
I did a last check and I fixed two things.
1. an image was not showing as I assumed the Ruffian Unit was an outlaw unit and not a peasant (which it actually is).
2. I have added a note indicating that the testers have access to three custom right-click menus for probing the AI behaviour.

I think I should have added that note earlier so that the missing image thing would have been revealed sooner.
Initially I speculated that the image not showing was caused by the crop IPF but the image path itself was mis-typed.

**Comment by knyghtmare** [comment]
Here is also my draft of the bloodlust MAI's wiki entry:
## Bloodlust MAI
The Bloodlust Micro AI leads a unit (or units) through behavior analogous to hyper aggressive or boss-class unit(s). The units wander in a prescribed area of the map, their "arena" zone, until an enemy unit comes within range. Note that these units do not actively chase down enemies when they enter the zone, but wander randomly through the arena zone until an enemy unit is encountered.
 If enemies are within range of the bloodlust-MAI unit, this unit attacks the weakest of them. If a kill is made, the unit resets its status and either roams the arena again (if no enemies are within range) or targets another of the weakest enemies if they are within attack range.

Note: this Micro AI is best optimized for dungeon bosses/minibosses where arena zone definitions are most prominent or for portraying the behavior of "berserker" units which are too frenzied to either see reason or cooperate with units belonging to the same side.

Is the draft good to go?

File: data/ai/micro_ais/cas/ca_bloodlust.lua
Comment (stevecotton) [inline data/ai/micro_ais/cas/ca_bloodlust.lua:63]: I've only just realised that `bloodlust_unit_status` is only set to `nil`. Unlike the Hunter CA's `hunting_status`, this variable doesn't seem to have a purpose.

It might be counting as a change for the purpose of checking whether the execution function has changed the game-state, but if it is then that would likely be a bug.

**Review by stevecotton — CHANGES_REQUESTED** [review]
> If enemies are within range of the bloodlust-MAI unit, this unit attacks the weakest of them. If a kill is made, the unit resets its status and either roams the arena again (if no enemies are within range) or targets another of the weakest enemies if they are within attack range.

Sorry for flip-flopping on the approval, but this paragraph made me notice something about the code. This description suggests that there's some logic of locking-on to the choice of target enemy, but AFAICS the only logic in the code that affects future turns is during the wandering cycle, when it selects an x,y pair to wander to.

**Comment by CelticMinstrel** [comment]
Since the AI was renamed, could you update the PR title?

**Comment by stevecotton** [comment]
Currently I think this is waiting on @knyghtmare to respond to the review about `bloodlust_unit_status`. No rush, this comment is just to keep track of what is waiting on what.

**Comment by Pentarctagon** [comment]
@knyghtmare any updates on this?

**Comment by knyghtmare** [comment]
> any updates on this?

I believe it was later decided that the MAI would be not necessary and a `return_home=true/false` bool key would be added to the hunter MAI instead which would give the same behaviour.

In any case, these Bloodlust MAI and Minions MAI will be coming packaged in NR-rework. The decision to add to core or not can come later.

**Comment by Pentarctagon** [comment]
Alright, so it sounds like this PR can be closed since it'll be superseded by the NR rework?

**Comment by Pentarctagon** [comment]
*pokes @knyghtmare*

**Comment by Pentarctagon** [comment]
Thanks.

**Pull request closed without merging**
